--- a/tools/xcutils/Makefile
+++ b/tools/xcutils/Makefile
@@ -24,7 +24,7 @@
 LDLIBS_xc_restore := $(LDLIBS_libxenctrl) $(LDLIBS_libxenguest)
 LDLIBS_xc_save    := $(LDLIBS_libxenctrl) $(LDLIBS_libxenguest) $(LDLIBS_libxenstore)
 LDLIBS_readnotes  := $(LDLIBS_libxenctrl) $(LDLIBS_libxenguest)
-LDLIBS_lsevtchn   := $(LDLIBS_libxenctrl)
+LDLIBS_lsevtchn   := $(LDLIBS_libxenctrl) $(call LDFLAGS_RPATH,../lib)
 
 .PHONY: all
 all: build
